src/v/kafka/server/handlers/alter_partition_reassignments.cc

```diff
@@ -0,0 +1,283 @@
+
+#include "kafka/server/handlers/alter_partition_reassignments.h"
+
+#include "cluster/errc.h"
+#include "cluster/metadata_cache.h"
+#include "cluster/topics_frontend.h"
+#include "cluster/types.h"
+#include "kafka/protocol/errors.h"
+#include "kafka/protocol/schemata/alter_partition_reassignments_request.h"
+#include "kafka/protocol/schemata/alter_partition_reassignments_response.h"
+#include "kafka/server/errors.h"
+#include "kafka/server/fwd.h"
+#include "kafka/server/quota_manager.h"
+#include "model/metadata.h"
+#include "model/namespace.h"
+#include "model/timeout_clock.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/sstring.hh>
+
+#include <absl/container/node_hash_map.h>
+#include <fmt/format.h>
+
+#include <algorithm>
+#include <chrono>
+#include <iterator>
+#include <system_error>
+#include <vector>
+
+namespace kafka {
+
+error_code cluster_errc_to_kafka_error(std::error_code ec) {
+    using clerrc = cluster::errc;
+
+    switch (static_cast<clerrc>(ec.value())) {
+    case clerrc::success:
+        return error_code::none;
+    case clerrc::topic_not_exists:
+        return error_code::unknown_topic_or_partition;
+    case clerrc::update_in_progress:
+        return error_code::reassignment_in_progress;
+    case clerrc::no_update_in_progress:
+        return error_code::no_reassignment_in_progress;
+    default:
+        return error_code::unknown_server_error;
+    }
+}
+
+template<>
+ss::future<response_ptr> alter_partition_reassignments_handler::handle(
+  request_context ctx, [[maybe_unused]] ss::smp_service_group ssg) {
+    alter_partition_reassignments_request request;
+    request.decode(ctx.reader(), ctx.header().version);
+    log_request(ctx.header(), request);
+    alter_partition_reassignments_response resp;
+
+    if (request.data.topics.empty()) {
+        co_return co_await ctx.respond(std::move(resp));
+    }
+
+    if (!ctx.authorized(
+          security::acl_operation::alter, security::default_cluster_name)) {
+        resp.data.error_code = error_code::cluster_authorization_failed;
+        resp.data.error_message = ss::sstring{
+          error_code_to_str(error_code::cluster_authorization_failed)};
+        co_return co_await ctx.respond(std::move(resp));
+    }
+
+    resp.data.responses.reserve(request.data.topics.size());
+    auto timeout_ms = request.data.timeout_ms;
+    kafka::error_code resp_error{error_code::none};
+
+    for (const reassignable_topic& topic : request.data.topics) {
+        auto topic_config = ctx.metadata_cache().get_topic_cfg(
+          model::topic_namespace{model::kafka_namespace, topic.name});
+        if (!topic_config) {
+            vlog(klog.error, "Unknown topic: {}", topic.name);
+            resp.data.error_code = error_code::unknown_topic_or_partition;
+            resp.data.error_message = ss::sstring{
+              error_code_to_str(error_code::unknown_topic_or_partition)};
+            co_return co_await ctx.respond(std::move(resp));
```

Comment:
Returning here will exit the loop and we will not process the other topics in the input list. We need to first perform validation on the input, moving the items that fail validation into the response list, then take the remaining items that have passed validation, process them and put the responses in the final response list. Then return the total list of responses to the client as one.

Reply:
finally got this figured out.